lib/scion.js

```diff
@@ -401,73 +420,67 @@
     /* begin ArraySet */
 
     /** @constructor */
-    function ArraySet(l) {
+    function ArraySet(l) {        
         l = l || [];
-        this.o = [];
-            
-        l.forEach(function(x){
-            this.add(x);
-        },this);
+        this.o = new Set(l);        
     }
 
     ArraySet.prototype = {
 
         add : function(x) {
-            if (!this.contains(x)) return this.o.push(x);
+            this.o.add(x);
         },
 
         remove : function(x) {
-            var i = this.o.indexOf(x);
-            if(i === -1){
-                return false;
-            }else{
-                this.o.splice(i, 1);
-            }
-            return true;
+            return this.o.delete(x);
         },
 
         union : function(l) {
-            l = l.iter ? l.iter() : l;
-            l.forEach(function(x){
-                this.add(x);
-            },this);
+            for (var v of l.o) {
+                this.o.add(v);
+            }
             return this;
         },
 
         difference : function(l) {
-            l = l.iter ? l.iter() : l;
-
-            l.forEach(function(x){
-                this.remove(x);
-            },this);
+            for (var v of l.o) {
+                this.o.delete(v);
+            }
             return this;
         },
 
         contains : function(x) {
-            return this.o.indexOf(x) > -1;
+            return this.o.has(x);
         },
 
         iter : function() {
-            return this.o;
+            return Array.from(this.o);
         },
 
         isEmpty : function() {
-            return !this.o.length;
+            return !this.o.size;
+        },
+
+        size: function() {
+            return this.o.size;
         },
 
         equals : function(s2) {
-            var l2 = s2.iter();
-            var l1 = this.o;
+            if (this.o.size !== s2.size()) {
+                return false;
+            }
 
-            return l1.every(function(x){
-                return l2.indexOf(x) > -1;
-            }) && l2.every(function(x){
-                return l1.indexOf(x) > -1;
-            });
+            for (var v of this.o) {
+                if (!s2.contains(v)) {
+                    return false;
+                }
+            }
+
+            return true;
         },
 
         toString : function() {
-            return "Set(" + this.o.toString() + ")";
+            return "Set(" + Array.from(this.o).toString() + ")";
         }
     };
```

Comment:
We might be able to phase out ArraySet and just use the native ES6 Set implementation directly.